Use `deformable_detr` kernel from the Hub

danieldk opened the pull request:
# What does this PR do?

Remove the `deformable_detr` kernel from `kernels/` and use the pre-built kernel from the Hub instead.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @LysandreJik 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @eustlb
- graph models: @clefourrier

Library:

- flax: @gante and @Rocketknight1
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @Rocketknight1
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

danieldk committed:
Use `deformable_detr` kernel from the Hub

Remove the `deformable_detr` kernel from `kernels/` and use the
pre-built kernel from the Hub instead.

ArthurZucker commented on setup.py:
```diff
@@ -432,6 +434,7 @@ def run(self):
 install_requires = [
     deps["filelock"],  # filesystem locks, e.g., to prevent parallel downloads
     deps["huggingface-hub"],
+    deps["kernels"],  # download kernels from the Hub
```
That's the only place I'd want to wait @LysandreJik's input on! IMO this is the longterm plan for sure, wondering if we are gonna do that slowly (first optional soft dep) or not!

ArthurZucker commented on src/transformers/integrations/hub_kernels.py:
```diff
@@ -0,0 +1,28 @@
+from typing import Dict, Union
```
just missing the licence!

danieldk replied:
Added :heavy_check_mark:.

ArthurZucker reviewed (approved):
Super super super nice! 🚀 let's just wait for @LysandreJik 's opinion on adding a new core dep!

danieldk committed:
Add license header

danieldk committed:
Add `kernels` as an extra `hub-kernels`

Also add it to `testing`, so that the kernel replacement gets tested
when using CUDA in CI.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_36853). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ArthurZucker commented on src/transformers/integrations/hub_kernels.py:
```diff
@@ -0,0 +1,73 @@
+
+    register_kernel_mapping(_KERNEL_MAPPING)
+
+except ImportError:
```
perfect!

ArthurZucker reviewed (approved):
Let's GO!!! 🤗

qubvel commented:
Hey @danieldk! Super nice to see kernels integrated! Is there a way to disable the kernel path dynamically? RT-DETR's `compile` full graph and `torch.export` are broken now, they should go through the torch path.

ArthurZucker commented:
Ah normally if cuda is not available it should fallback to noraml forwar + if not available should do nothing

ArthurZucker commented:
(@qubvel export can be fast test no? sorry that it broke)

qubvel commented:
it's too slow, even though it's on a small model

qubvel commented:
we discussed with @danieldk in Slack, the idea is to have a way to control which path to go through, smth like

```python
if self.disable_custom_kernels or is_torchdynamo_compiling():
    with use_kernel_mapping({}, inherit_mapping=False): # inherit_mapping would be a new argument
        output = self.attn(
            value,
            spatial_shapes,
            spatial_shapes_list,
            level_start_index,
            sampling_locations,
            attention_weights,
            self.im2col_step,
        )
else:
    ...
```

ArthurZucker commented:
Or we register the use kernel with torch API to allow compile ?

qubvel commented:
Not aware of this but it can be an option, yes. Anyway it would be nice to be able to disable it manually in case there are any issues with the kernel + understand which path is executed (e.g. for RT-DETR to export for CoreML we need 5D tensors instead of 6D passed into the kernel, so I refactored pytorch path of deformable attention in this [PR](https://github.com/huggingface/transformers/pull/36514))
